server/src/main/java/org/opensearch/index/codec/customcodecs/ZstdCompressionMode.java

```diff
@@ -170,32 +172,33 @@ public void decompress(DataInput in, int originalLength, int offset, int length,
 
                 // decompress dictionary first
                 doDecompress(in, dctx, bytes, dictLength);
-
-                dctx.loadDict(new ZstdDictDecompress(bytes.bytes, 0, dictLength));
-
-                int offsetInBlock = dictLength;
-                int offsetInBytesRef = offset;
-
-                // Skip unneeded blocks
-                while (offsetInBlock + blockLength < offset) {
-                    final int compressedLength = in.readVInt();
-                    in.skipBytes(compressedLength);
-                    offsetInBlock += blockLength;
-                    offsetInBytesRef -= blockLength;
+                try (ZstdDictDecompress dictDecompress = new ZstdDictDecompress(bytes.bytes, 0, dictLength)) {
```

Comment:
Maybe we should use the other ctor? And do this with direct memory instead of using heap?

```
ZstdDictDecompress dictDecompress(ByteBuffer.wrap(bytes.bytes, 0, dictLength))
```

Reply:
The leak is within the native memory and didn't show any issues within the JVM heap. The allocations in native side of code are not freed if this was not closed, and hence lead to the leak.